### ISSUE TITLE
[ARM 32-bit] Unable to start debugging with an android physical device (Hot Reload)

### ISSUE THREAD

**Issue by ArtefactoDev** [dotnet/android#9573]
### Android framework version

net9.0-android

### Affected platform version

VS 17.12.2, SDK 9.0.100, Windows 11 24H2

### Description

The system compiles the code and sends it to the Android phone, however, after generating the .apk package in the system correctly, Visual Studio does not allow me to debug it.

I can open the generated package on my cell phone without errors, but I cannot debug it using 'hot reload' in any way.

Cell phone model: Motorola G(8)
Android Version: Android 10

Detail, on the same machine I have another application developed in Xamarin for the same cell phone, and I can debug it normally.

### Steps to Reproduce

Video Error:
https://github.com/user-attachments/assets/4260e896-6c6a-4fd8-9101-87aff3ee9893

1 - Start Visual Studio
2 - Create a New Project
3 - .NET MAUI App (name: ExemploAndroid)

4 - Net 9
('include sample content' NOT selected)

Change the file 'ExemploAndroid.csproj' to:

```c#

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>net9.0-android;net9.0-ios;net9.0-maccatalyst</TargetFrameworks>
    <TargetFrameworks Condition="$([MSBuild]::IsOSPlatform('windows'))">$(TargetFrameworks);net9.0-windows10.0.19041.0</TargetFrameworks>
    <!-- Uncomment to also build the tizen app. You will need to install tizen by following this: https://github.com/Samsung/Tizen.NET -->
    <!-- <TargetFrameworks>$(TargetFrameworks);net9.0-tizen</TargetFrameworks> -->

    <!-- Note for MacCatalyst:
		The default runtime is maccatalyst-x64, except in Release config, in which case the default is maccatalyst-x64;maccatalyst-arm64.
		When specifying both architectures, use the plural <RuntimeIdentifiers> instead of the singular <RuntimeIdentifier>.
		The Mac App Store will NOT accept apps with ONLY maccatalyst-arm64 indicated;
		either BOTH runtimes must be indicated or ONLY macatalyst-x64. -->
    <!-- For example: <RuntimeIdentifiers>maccatalyst-x64;maccatalyst-arm64</RuntimeIdentifiers> -->

    <OutputType>Exe</OutputType>
    <RootNamespace>ExemploAndroid</RootNamespace>
    <UseMaui>true</UseMaui>
    <SingleProject>true</SingleProject>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>

    <!-- Display name -->
    <ApplicationTitle>ExemploAndroid</ApplicationTitle>

    <!-- App Identifier -->
    <ApplicationId>com.companyname.exemploandroid</ApplicationId>

    <!-- Versions -->
    <ApplicationDisplayVersion>1.0</ApplicationDisplayVersion>
    <ApplicationVersion>1</ApplicationVersion>

    <!-- To develop, package, and publish an app to the Microsoft Store, see: https://aka.ms/MauiTemplateUnpackaged -->
    <WindowsPackageType>None</WindowsPackageType>
    <SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'">15.0</SupportedOSPlatformVersion>
    <SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'maccatalyst'">15.0</SupportedOSPlatformVersion>
    <SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">28.0</SupportedOSPlatformVersion>
    <RuntimeIdentifiers Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">android-arm;android-arm64;android-x86;android-x64;</RuntimeIdentifiers>
    <SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">10.0.17763.0</SupportedOSPlatformVersion>
    <TargetPlatformMinVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">10.0.17763.0</TargetPlatformMinVersion>
    <SupportedOSPlatformVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'tizen'">6.5</SupportedOSPlatformVersion>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net9.0-android|AnyCPU'">
    <EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>
    <!--<AndroidSupportedAbis>true</AndroidSupportedAbis>-->
    <!-- <AndroidCreatePackagePerAbi>True</AndroidCreatePackagePerAbi>-->
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|net9.0-android|AnyCPU'">
    <EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>
    <!-- <AndroidCreatePackagePerAbi>True</AndroidCreatePackagePerAbi>-->
  </PropertyGroup>


  <ItemGroup>
    <!-- App Icon -->
    <MauiIcon Include="Resources\AppIcon\appicon.svg" ForegroundFile="Resources\AppIcon\appiconfg.svg" Color="#512BD4" />

    <!-- Splash Screen -->
    <MauiSplashScreen Include="Resources\Splash\splash.svg" Color="#512BD4" BaseSize="128,128" />

    <!-- Images -->
    <MauiImage Include="Resources\Images\*" />
    <MauiImage Update="Resources\Images\dotnet_bot.png" Resize="True" BaseSize="300,185" />

    <!-- Custom Fonts -->
    <MauiFont Include="Resources\Fonts\*" />

    <!-- Raw Assets (also remove the "Resources\Raw" prefix) -->
    <MauiAsset Include="Resources\Raw\**" LogicalName="%(RecursiveDir)%(Filename)%(Extension)" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Maui.Controls" Version="9.0.10" />
    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="9.0.0" />
  </ItemGroup>

</Project>

``` 

Plug in your cell phone (already configured in developer mode), choose to debug and after deployment, Visual Studio stops debugging.

### Did you find any workaround?

unfortunately not.

### Relevant log output

```shell
Build started at 21:20...
1>------ Build started: Project: ExemploAndroid, Configuration: Debug Any CPU ------
1>Skipping analyzers to speed up the build. You can execute 'Build' or 'Rebuild' command to run analyzers.
1>Including assemblies for Hot Reload support
1>ExemploAndroid -> C:\Users\workc\source\repos\ExemploAndroid\bin\Debug\net9.0-android\ExemploAndroid.dll
2>------ Deploy started: Project: ExemploAndroid, Configuration: Debug Any CPU ------
2>Starting deployment to Motorola moto g(8) play ...
2>Deploying to Motorola moto g(8) play ...
Build started.
Project "ExemploAndroid.csproj" (Install target(s)):
There is no application host available for the specified RuntimeIdentifier 'android-arm'.
There is no application host available for the specified RuntimeIdentifier 'android-arm64'.
There is no application host available for the specified RuntimeIdentifier 'android-x86'.
There is no application host available for the specified RuntimeIdentifier 'android-x64'.
Found Java SDK version 17.0.12.
MonoAndroid Tools: c:\program files\dotnet\packs\Microsoft.Android.Sdk.Windows\35.0.7\tools\
Android Platform API level: 35
TargetFrameworkVersion: v9.0
Android NDK: 
Android SDK: C:\Program Files (x86)\Android\android-sdk\
Android SDK Build Tools: C:\Program Files (x86)\Android\android-sdk\build-tools\35.0.0\
Java SDK: C:\Program Files (x86)\Android\openjdk\jdk-17.0.12\Application Java class: android.app.Application
_OuterIntermediateOutputPath: IntermediateOutputPath: obj\Debuget9.0-android\
Skipping target "ProcessMauiSplashScreens" because all output files are up-to-date with respect to the input files.
Skipping target "ProcessMauiFonts" because all output files are up-to-date with respect to the input files.
Skipping target "ResizetizeImages" because all output files are up-to-date with respect to the input files.
Skipping target "_ResolveLibraryProjectImports" because all output files are up-to-date with respect to the input files.
Skipping target "_BuildLibraryImportsCache" because all output files are up-to-date with respect to the input files.
Skipping target "_GenerateResourceCaseMap" because all output files are up-to-date with respect to the input files.
Skipping target "_GenerateResourceDesignerIntermediateClass" because all output files are up-to-date with respect to the input files.
Skipping target "_GenerateLayoutBindings" because it has no outputs.
Skipping target "_ConvertResourcesCases" because all output files are up-to-date with respect to the input files.
Skipping target "_CompileResources" because all output files are up-to-date with respect to the input files.
Skipping target "_PrepareUpdateAndroidResgen" because all output files are up-to-date with respect to the input files.
Skipping target "_UpdateAndroidResgen" because all output files are up-to-date with respect to the input files.
Project "ExemploAndroid.csproj" (_ComputeFilesToPublishForRuntimeIdentifiers target(s)):
There is no application host available for the specified RuntimeIdentifier 'android-arm'.
There is no application host available for the specified RuntimeIdentifier 'android-arm64'.
There is no application host available for the specified RuntimeIdentifier 'android-x86'.
There is no application host available for the specified RuntimeIdentifier 'android-x64'.
_OuterIntermediateOutputPath: obj\Debuget9.0-android\IntermediateOutputPath: obj\Debuget9.0-android\android-arm\
Done building project "ExemploAndroid.csproj".
Project "ExemploAndroid.csproj" (_ComputeFilesToPublishForRuntimeIdentifiers target(s)):
There is no application host available for the specified RuntimeIdentifier 'android-arm'.
There is no application host available for the specified RuntimeIdentifier 'android-arm64'.
There is no application host available for the specified RuntimeIdentifier 'android-x86'.
There is no application host available for the specified RuntimeIdentifier 'android-x64'.
_OuterIntermediateOutputPath: obj\Debuget9.0-android\IntermediateOutputPath: obj\Debuget9.0-android\android-arm64\Done building project "ExemploAndroid.csproj".
Project "ExemploAndroid.csproj" (_ComputeFilesToPublishForRuntimeIdentifiers target(s)):
There is no application host available for the specified RuntimeIdentifier 'android-arm'.
There is no application host available for the specified RuntimeIdentifier 'android-arm64'.
There is no application host available for the specified RuntimeIdentifier 'android-x86'.
There is no application host available for the specified RuntimeIdentifier 'android-x64'.
_OuterIntermediateOutputPath: obj\Debuget9.0-android\IntermediateOutputPath: obj\Debuget9.0-android\android-x86\Done building project "ExemploAndroid.csproj".
Project "ExemploAndroid.csproj" (_ComputeFilesToPublishForRuntimeIdentifiers target(s)):
There is no application host available for the specified RuntimeIdentifier 'android-arm'.
There is no application host available for the specified RuntimeIdentifier 'android-arm64'.
There is no application host available for the specified RuntimeIdentifier 'android-x86'.
There is no application host available for the specified RuntimeIdentifier 'android-x64'.
_OuterIntermediateOutputPath: obj\Debuget9.0-android\IntermediateOutputPath: obj\Debuget9.0-android\android-x64\Done building project "ExemploAndroid.csproj".
Skipping target "_LinkAssembliesNoShrink" because all output files are up-to-date with respect to the input files.
Skipping target "_GenerateJavaStubs" because all output files are up-to-date with respect to the input files.
Skipping target "_ManifestMerger" because all output files are up-to-date with respect to the input files.
Skipping target "_ConvertCustomView" because all output files are up-to-date with respect to the input files.
Skipping target "_AddStaticResources" because all output files are up-to-date with respect to the input files.
Skipping target "_GeneratePackageManagerJava" because all output files are up-to-date with respect to the input files.
Skipping target "_GenerateAndroidAssetsDir" because all output files are up-to-date with respect to the input files.
Skipping target "_PrepareCreateBaseApk" because all output files are up-to-date with respect to the input files.
Skipping target "_CreateBaseApk" because all output files are up-to-date with respect to the input files.
Skipping target "_CompileJava" because all output files are up-to-date with respect to the input files.
Skipping target "_CompileNativeAssemblySources" because all output files are up-to-date with respect to the input files.
Skipping target "_CreateApplicationSharedLibraries" because all output files are up-to-date with respect to the input files.
Skipping target "_CompileToDalvik" because all output files are up-to-date with respect to the input files.
Deleting file "bin\Debuget9.0-android\com.companyname.exemploandroid.apk".
Copying file from "C:\Users\workc\source\repos\ExemploAndroid\obj\Debug
et9.0-android\android\bin\com.companyname.exemploandroid.apk" to "C:\Users\workc\source\repos\ExemploAndroid\bin\Debug
et9.0-android\com.companyname.exemploandroid.apk".
Touching "obj\Debug
et9.0-android\android_debug_keystore.flag".
Deleting file "C:\Users\workc\source\repos\ExemploAndroid\bin\Debug
et9.0-android\com.companyname.exemploandroid-Signed.apk".
C:\Program Files (x86)\Android\android-sdk\build-tools\35.0.0\zipalign.exe -p 16 "obj\Debug
et9.0-android\android\bin\com.companyname.exemploandroid.apk" "bin\Debug
et9.0-android\\com.companyname.exemploandroid-Signed.apk" 
C:\Program Files (x86)\Android\openjdk\jdk-17.0.12\bin\java.exe -jar "C:\Program Files (x86)\Android\android-sdk\build-tools\35.0.0\lib\apksigner.jar" sign --ks "C:\Users\workc\AppData\Local\Xamarin\Mono for Android\debug.keystore" --ks-pass pass:android --ks-key-alias androiddebugkey --key-pass pass:android --min-sdk-version 28 --max-sdk-version 35  bin\Debug
et9.0-android\com.companyname.exemploandroid-Signed.apk 
Signed android package 'bin\Debug
et9.0-android\com.companyname.exemploandroid-Signed.apk'
Using cached value from RegisterTaskObject
Found device: 0062963070
Forwarding debugger port 8805
Detecting existing process
> am start -a "android.intent.action.MAIN" -c "android.intent.category.LAUNCHER" -n "com.companyname.exemploandroid/crc64358cda76bdc6f75f.MainActivity"
> Starting: Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] cmp=com.companyname.exemploandroid/crc64358cda76bdc6f75f.MainActivity }
[.exemploandroi] Late-enabling -Xcheck:jni
[.exemploandroi] Unknown bits set in runtime_flags: 0x8000
[DOTNET] AndroidCryptoNative_InitLibraryOnLoad: jint AndroidCryptoNative_InitLibraryOnLoad(JavaVM *, void *) in /__w/1/s/src/native/libs/System.Security.Cryptography.Native.Android/pal_jni.c
[DOTNET] GetOptionalClassGRef: optional class com/android/org/conscrypt/OpenSSLEngineImpl was not found
[monodroid] Failed to create directory '/data/user/0/com.companyname.exemploandroid/files/.__override__/armeabi-v7a'. No such file or directory
[monodroid] Creating public update directory: `/data/user/0/com.companyname.exemploandroid/files/.__override__/armeabi-v7a`
[.exemploandroi] Attempt to remove non-JNI local reference, dumping thread
[monodroid-debug] Trying to initialize the debugger with options: --debugger-agent=transport=dt_socket,loglevel=0,address=127.0.0.1:8805,server=y,embedding=1,timeout=8
[monodroid-assembly] Unable to open app library directory '/data/user/0/com.companyname.exemploandroid/files/.__override__/armeabi-v7a': No such file or directory
[mono] Listening on 127.0.0.1:8805 (timeout=8 ms)...
[mono] debugger-agent: Timed out waiting to connect.
```

**Comment by ArtefactoDev** [comment]
>This phrasing is confusing, are you trying to debug or use hot reload? The error looks like the debugger is not connecting.

I can't debug when I try to use a physical Android device.

>Does the default template with no changes work? 

No work.
if you try to generate a default template (same Steps to Reproduce ) and try to debug (without any modification) using a physical Android device you will receive the following error:

`ADB0020: Mono.AndroidTools.IncompatibleCpuAbiException: The package does not support the CPU architecture of this device`

Shouldn't the template, without any modifications, work without errors?


>What change did you make that "breaks" things?

Basically, trying to get around the 'ADB0020' error, I only added these lines:

`
 <RuntimeIdentifiers Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">android-arm;android-arm64;android-x86;android-x64;</RuntimeIdentifiers>
`

`
   <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net9.0-android|AnyCPU'">
    <EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>
  </PropertyGroup>
`

**Comment by jonathanpeppers** [comment]
> if you try to generate a default template (same Steps to Reproduce ) and try to debug (without any modification) using a physical Android device you will receive the following error:

I do this daily, and it is working for me. What device are you using? Can you share a `.binlog` of the failure? https://aka.ms/binlog

> ADB0020: Mono.AndroidTools.IncompatibleCpuAbiException: The package does not support the CPU architecture of this device

`ADB0020` is a completely different issue than debugging or hot reload. It means the deploy failed.

Is the device a 32-bit device? If so, you would need to list `android-arm` as a `RuntimeIdentifier` as you tried:
```xml
<RuntimeIdentifiers Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">android-arm;android-arm64;android-x64;</RuntimeIdentifiers>
```

A `.binlog` would allow us to investigate further, thanks!

**Comment by ArtefactoDev** [comment]
>  do this daily, and it is working for me. What device are you using? 

Not for me, even using the default template without any changes
It works normally if I use the Android emulator, but not with the physical device.

Are you trying an emulator or are you trying to use a physical device for this test? because in the emulator it works normally.
Physical device: Motorola G(8) (Android 10)


>Can you share a .binlog of the failure? https://aka.ms/binlog

I attributed the log to open ticket to try to help.[EngineTrace_4472.txt](https://github.com/user-attachments/files/17998364/EngineTrace_4472.txt)
[EngineTrace_12004.txt](https://github.com/user-attachments/files/17998374/EngineTrace_12004.txt)
[EngineTrace_23944.txt](https://github.com/user-attachments/files/17998366/EngineTrace_23944.txt)
[EngineTrace_26780.txt](https://github.com/user-attachments/files/17998372/EngineTrace_26780.txt)
[EngineTrace_27524.txt](https://github.com/user-attachments/files/17998371/EngineTrace_27524.txt)
[EngineTrace_28940.txt](https://github.com/user-attachments/files/17998370/EngineTrace_28940.txt)
[EngineTrace_29052.txt](https://github.com/user-attachments/files/17998379/EngineTrace_29052.txt)
[MSBuild_CommTrace_PID_4472.txt](https://github.com/user-attachments/files/17998373/MSBuild_CommTrace_PID_4472.txt)
[MSBuild_CommTrace_PID_12004.txt](https://github.com/user-attachments/files/17998367/MSBuild_CommTrace_PID_12004.txt)
[MSBuild_CommTrace_PID_23944.txt](https://github.com/user-attachments/files/17998378/MSBuild_CommTrace_PID_23944.txt)
[MSBuild_CommTrace_PID_26780.txt](https://github.com/user-attachments/files/17998368/MSBuild_CommTrace_PID_26780.txt)
[MSBuild_CommTrace_PID_27524.txt](https://github.com/user-attachments/files/17998376/MSBuild_CommTrace_PID_27524.txt)
[MSBuild_CommTrace_PID_28940.txt](https://github.com/user-attachments/files/17998375/MSBuild_CommTrace_PID_28940.txt)
[MSBuild_CommTrace_PID_29052.txt](https://github.com/user-attachments/files/17998369/MSBuild_CommTrace_PID_29052.txt)
[SchedulerState_12004.txt](https://github.com/user-attachments/files/17998377/SchedulerState_12004.txt)
[SchedulerTrace_12004.txt](https://github.com/user-attachments/files/17998365/SchedulerTrace_12004.txt)


> ADB0020 is a completely different issue than debugging or hot reload. It means the deploy failed.


I know, I explained what happens when you try to debug a physical device using the default template, without any changes.
I just find it strange that the default template without any changes, but with .Net 8 (templante default) works without any problems...


> Is the device a 32-bit device? If so, you would need to list android-arm as a RuntimeIdentifier as you tried: (...)

I added all the code from the 'csproj' project to the ticket.
I tried what you said, but it still didn't work.

Maybe the problem is with arm.. But, don't you agree that this should already come by default in the default template, if we consider that more than 90% of the time we will come across ARM devices on Android?

Thank you very much for the excellent help, I'm sure the team will solve it! Because you are excellent!

**Comment by jonathanpeppers** [comment]
The logs above are not `.binlog` files. Were you able to collect those?

> 90% of the time we will come across ARM devices on Android?

The default is `android-arm64;android-x64` in .NET 9, you would have a very old Android device for it to be `arm` and 32-bit.

**Comment by ArtefactoDev** [comment]
>The logs above are not .binlog files. Were you able to collect those?

[binlog.zip](https://github.com/user-attachments/files/18000481/binlog.zip)

> The default is android-arm64;android-x64 in .NET 9, you would have a very old Android device for it to be arm and 32-bit.

It's a possibility.
Is there a way I can get around this to test?

**Comment by dellis1972** [comment]
You already mentioned the solution for this which is to add `android-arm` to the list of `RuntimeIdentifiers`. 

```
<RuntimeIdentifiers Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">android-arm;android-arm64;android-x86;android-x64;</RuntimeIdentifiers>
```

**Comment by ArtefactoDev** [comment]
Device Android:
Motorola g(8) play (SKU: XT2015-2)
Android 10 (api Level 20)
Processor: MediaTek Helio P70M
Kernel Arquitecture: armv7l 


Okay, I cleared everything.
I recreated everything from scratch, to test.

1 - I created a default template without any changes to the '.Net Maui App'.

2 - If I try to debug on the Android device), it will give an error:
`ADB0020: Mono.AndroidTools.IncompatibleCpuAbiException: The package does not support the CPU architecture of this device. ...`

3 - I add the line you passed:
`<RuntimeIdentifiers Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">android-arm;android-arm64;android-x86;android-x64;</RuntimeIdentifiers>`

4 - Return the error
`NETSDK1047: Assets file 'C:\Users\workc\source\repos\MauiApp1\obj\project.assets.json' doesn't have a target for 'net9.0-android/android-arm'. Ensure that restore has run and that you have included 'net9.0-android' in the TargetFrameworks for your project. You may also need to include 'android-arm' in your project's RuntimeIdentifiers.`

5 - Add 'project.assets.json' to help

**Comment by jonathanpeppers** [comment]
> NETSDK1047: Assets file 'C:\Users\workc\source\repos\MauiApp1\obj\project.assets.json' doesn't have a target for 'net9.0-android/android-arm'. Ensure that restore has run and that you have included 'net9.0-android' in the TargetFrameworks for your project. You may also need to include 'android-arm' in your project's RuntimeIdentifiers.

This feels like an IDE/Visual Studio/project system bug, but you may have to close and reopen the project to force it to run NuGet restore again. You might also try `dotnet restore` at the command-line

**Comment by ArtefactoDev** [comment]
I try, but same error.
Is there anything else I can do?

**Comment by ArtefactoDev** [comment]
I try add two lines:

` <PropertyGroup>`
 `  <RuntimeIdentifiers Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">android-arm;android-arm64;android-x86;android-x64;</RuntimeIdentifiers>`
  ` <Platforms>AnyCPU;ARM64;ARM32</Platforms>`
 `</PropertyGroup>
`


By choosing arm32, I can generate the apk and add it to the device, but the error that started it all comes back (not able to debug):
`[monodroid-debug] Trying to initialize the debugger with options: --debugger-agent=transport=dt_socket,loglevel=0,address=127.0.0.1:8866,server=y,embedding=1,timeout=8`
`[mono] Listening on 127.0.0.1:8866 (timeout=8 ms)...`
`[mono] debugger-agent: Timed out waiting to connect.`

I feel like I'm in a loop.

**Comment by ArtefactoDev** [comment]
I formatted my machine. I only installed Visual Studio.

I chose the "MAUI App" model and without any modifications I tried to generate the debug on the physical device and I had exactly the same error.

I believe there must be a problem with Visual Studio generation.

**Comment by ArtefactoDev** [comment]
Good afternoon everyone, any news?

**Comment by jonathanpeppers** [comment]
If you are having trouble with a specific device:
```
Cell phone model: Motorola G(8)
Android Version: Android 10
armeabi-v7a / 32 bit arm
```

Does the emulator work just fine or a different device?

We would prioritize this based on how many people are using 32-bit arm devices. Or if it happens to every Motorola device, we would want to investigate.

**Comment by jonathanpeppers** [comment]
> You already mentioned the solution for this which is to add `android-arm` to the list of `RuntimeIdentifiers`.
> 
> ```
> <RuntimeIdentifiers Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">android-arm;android-arm64;android-x86;android-x64;</RuntimeIdentifiers>
> ```

@WilliamWatterson86 this is what was working for us. For a non-MAUI app, you wouldn't need the `Condition` at all.

**Comment by WilliamWatterson86** [comment]
@jonathanpeppers thanks for getting back to me.  

I am using a Samsung Galaxy A13 device (armeabi-v7a).  I am getting the same issues as @ArtefactoDev.  Blank template app does not deploy to my devicewithout adding in `<RuntimeIdentifiers Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'android'">android-arm;android-arm64;android-x86;android-x64;</RuntimeIdentifiers>`.

When the app deploys, it launches but crashes on splash screen.  I have checked logcat logs and can see "mono    : debugger-agent: Timed out waiting to connect." just before the crash.

Launching the app on phone without debugging works ok.  I checked on a arm64-v8a and it works - so seems to be something specific to armeabi-v7a.

A blank .net8 template app runs ok on this device.

**Comment by dellis1972** [comment]
If adding the armeabi-v7a to `RuntimeIdentifiers` fixes the app starting then that part is by design. We removed the 32bit platforms from the defaults of RuntimeIdentifiers on purpose given the very small number of those devices which are now in circulation. If you want to support 32bit it is expected that you will modify the RuntimeIdentifiers accordingly. 

the debugger not working is a different problem and probably a runtime issue. Have you tried turning Hot Reload off? to test? (I'm not sure how to do that but it can be done). We might need to move this issue to the runtime team to take a look.

**Comment by WilliamWatterson86** [comment]
@dellis1972 - yep I tried switching off hot reload, unfortunately this did not help.

Adding the line in def does make the app deploy.  This is fine.

The only other thing I see in the logs regarding runtime is:

There is no application host available for the specified RuntimeIdentifier 'android-arm'.
There is no application host available for the specified RuntimeIdentifier 'android-arm64'.
There is no application host available for the specified RuntimeIdentifier 'android-x86'.
There is no application host available for the specified RuntimeIdentifier 'android-x64'.

Maybe this is something?

I tried the preview version of Visual Studio too just to check.  It also has the same issue.

**Comment by dellis1972** [comment]
I don't know, we'd need someone from the runtime side to take a look. 
But it would help if you could post a full `logcat` of the issue. 

So from an Adb command prompt run `adb logcat -c`. Then run `adb logcat > log.txt` and then start your debug session. Once the issue is replicated, use `Ctrl+C` to exit the logcat process and then attach the log to this issue.

**Comment by WilliamWatterson86** [comment]
Thanks for getting back to me.  

I have attached the adb logs as requested.  My app is called MauiApp3

[logcat_maui.txt](https://github.com/user-attachments/files/18963724/logcat_maui.txt)

**Comment by ArtefactoDev** [comment]
> [@ArtefactoDev](https://github.com/ArtefactoDev) did you find any workaround for your device?


unfortunately not. No other way worked.
I tried several things as I described and none of them worked.

**Comment by dellis1972** [comment]
So I think I know what is causing the issue . In the log provided I see the following

```
monodroid-debug: Trying to initialize the debugger with options: --debugger-agent=transport=dt_socket,loglevel=0,address=127.0.0.1:8848,server=y,embedding=1,timeout=3
```

If you look at the line carefully you will see `timeout=3` which is 3 ms! 
I would expect to see something like `timeout=1716578797`. Not sure why you are getting a `3` there, but the debugger is basically exiting before the VS debugger has had a chance to attach to it. 

Can you run this on the device which has the issue and post the results

```
adb shell date +%s
```

**Comment by WilliamWatterson86** [comment]
@dellis1972 - here is the result I get when running this command on my device 1740575082

**Comment by WilliamWatterson86** [comment]
@dellis1972 I tried on an x64 device and get the log
02-26 13:46:52.079	Samsung SM-G973F	Warning	27206	monodroid-debug	Trying to initialize the debugger with options: --debugger-agent=transport=dt_socket,loglevel=0,address=127.0.0.1:8833,server=y,embedding=1,timeout=1740577641

and on x86 yes it shows this every time:
02-26 13:52:12.358	Samsung SM-A137F	Warning	29331	monodroid-debug	Trying to initialize the debugger with options: --debugger-agent=transport=dt_socket,loglevel=0,address=127.0.0.1:8834,server=y,embedding=1,timeout=3

**Comment by dellis1972** [comment]
@WilliamWatterson86 can you find out what the value for   `debug.mono.extra` is during a failed connection?

```
adb shell getprop debug.mono.extra
```

We are trying to figure out where the issue is. I suspect the runtime reading the timeout into a variable which is overflowing. But we need to know what the property is first.

**Comment by dellis1972** [comment]
@WilliamWatterson86 

have to tried this on other devices or emulators? (32 bit , 64 bit). 
Do you get the same results? Or is this a specific arm 32 bit device?

**Comment by WilliamWatterson86** [comment]
x64 bit devces seem to be ok, I have tried a couple of these.

I only have one x86 Samsung A13 device.  So I havent been able to try it on other x86 devices so far.

**Comment by dellis1972** [comment]
@WilliamWatterson86 

do you have android-arm in your `RuntimeIdentifiers` in your csproj?

**Comment by ArtefactoDev** [comment]
Wow, beautiful work.
Please when will this fix be available to use?

**Comment by dellis1972** [comment]
If we can get [Fix Debugger timeout on 32 bit devices for .net 9 #9880](https://github.com/dotnet/android/pull/9880) in, then it will be in the next .net 9 Release which is next month I think.
